olive/passes/onnx/add_metadata.py

```diff
@@ -0,0 +1,153 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+# --------------------------------------------------------------------------
+
+import logging
+from pathlib import Path
+
+import onnx
+
+from olive.hardware.accelerator import AcceleratorSpec
+from olive.model import ONNXModelHandler
+from olive.model.utils import resolve_onnx_path
+from olive.passes import Pass
+from olive.passes.onnx.common import model_proto_to_file, resave_model
+from olive.passes.pass_config import BasePassConfig, PassConfigParam
+
+logger = logging.getLogger(__name__)
+
+
+class AddOliveMetadata(Pass):
+    """Adds Olive-specific metadata to an ONNX model."""
+
+    @classmethod
+    def _default_config(cls, accelerator_spec: AcceleratorSpec) -> dict[str, PassConfigParam]:
+        return {
+            "graph_name": PassConfigParam(
+                type_=str,
+                required=True,
+                description="Custom graph name for the ONNX model. This field is required.",
+            ),
+            "custom_metadata": PassConfigParam(
+                type_=dict,
+                default_value={},
+                description="Custom metadata key-value pairs to add to the model",
+            ),
+            "add_olive_version": PassConfigParam(
+                type_=bool,
+                default_value=True,
+                description="Add Olive version to metadata",
+            ),
+            "add_optimization_info": PassConfigParam(
+                type_=bool,
+                default_value=True,
+                description="Add optimization information to metadata",
+            ),
+        }
+
+    def _add_metadata(self, onnx_model: onnx.ModelProto, metadata: dict[str, str]) -> onnx.ModelProto:
+        """Add metadata to the ONNX model."""
+        # Get existing metadata
+        existing_metadata = {entry.key: entry.value for entry in onnx_model.metadata_props}
+
+        # Update with new metadata (this will overwrite existing keys)
+        existing_metadata.update(metadata)
+
+        # Set the updated metadata
+        onnx.helper.set_model_props(onnx_model, existing_metadata)
+
+        return onnx_model
+
+    def _set_graph_name(self, onnx_model: onnx.ModelProto, graph_name: str) -> onnx.ModelProto:
+        """Set the ONNX graph name."""
+        onnx_model.graph.name = graph_name
+        return onnx_model
+
+    def _generate_olive_metadata(self, model: ONNXModelHandler, config: BasePassConfig) -> dict[str, str]:
+        """Generate Olive-specific metadata."""
+        metadata = {}
+
+        # Add custom metadata from config
+        if config.custom_metadata:
+            metadata.update({str(k): str(v) for k, v in config.custom_metadata.items()})
+
+        # Add Olive version
+        if config.add_olive_version:
```

Comment:
I'm thinking adding this to model anyway and no need a config param for this. Any case we don't want to add version here?